### ISSUE TITLE
Usage of Kconfig

### ISSUE THREAD

**Issue by errordeveloper** [kubernetes/kubernetes-anywhere#128]
There is a number of nice things about Kconfig.

If we would like to stick with Kconfig, it'd be nicer/easier to make it part of the build process, but it will require compiler installed. We probably could ship a container with all the dependencies.

An alternative could be to use Python, as it's very popular and easy to understand, it also doesn't require user to compile anything etc.

I've had a quick look, and there exists a usable [Kconfig parser](https://github.com/ulfalizer/Kconfiglib). We could potentially consider triggering jsonnet run from Python directly as well.

**Comment by mikedanese** [comment]
conf is currently precompiled and pulled down as needed by make:

https://github.com/kubernetes/kubernetes-anywhere/blob/master/Makefile#L25

We can compile it for mac and linux and publish prebuilt binaries as well as offering a docker image with all the depenedencies. That would get us out of the game of maintaining a bunch of scripts. I'd like to see how far we can go with little or no scripting.

**Comment by errordeveloper** [comment]
@mikedanese sure, I was just saving some observation I've made out of curiosity.

**Comment by errordeveloper** [comment]
I think providing a container would a good start. I'll create and issue and start working on that.

**Comment by colemickens** [comment]
I have one in the Azure PR. At least I think so based on that description.

**Comment by errordeveloper** [comment]
@colemickens 👍

**Comment by errordeveloper** [comment]
@colemickens [this](https://github.com/colemickens/kube-deploy/blob/19bf5e3f858e6b2f11986a12676e5807dc7c9ac4/min-turnup/Dockerfile) looks like exactly what we are talking about, please make a PR with just that dockerfile, or I copy it across myself...

**Comment by colemickens** [comment]
Okay, will send shortly.

On Thu, Jun 23, 2016 at 2:07 PM, Ilya Dmitrichenko <notifications@github.com

**Comment by colemickens** [comment]
https://github.com/kubernetes/kubernetes-anywhere/pull/147

**Comment by colemickens** [comment]
Do note, the file is slightly different. I couldn't get [this in the Makefile](https://github.com/kubernetes/kubernetes-anywhere/blob/1747396c9445ba4cea51633f660dacc661ac7694/Makefile#L3) to play nice with the `buildpack-deps` base Docker image so I left it as the ubuntu base.

**Comment by mikedanese** [comment]
Ok, Kconfig is annoying, kubernetes-anywhere developers have reported that they feel like they are working around it. We need a way to allow users to interactively discover configuration options both initially and as configuration options change and are added. We need to output the config choices as json. These options often have suboptions and dependencies so it would be nice to use a tool that understands this. Kconfig provides this but there are probably other options here. I'm open to suggestions. A quick google search doesn't turn up much. I'm open to suggestions here.

I like the idea of building this configuration object from the command line but a easy webapp that outputs the json config would also be very cool. Is there a way we could have both?

cc @madhusudancs @lukemarsden @mansoorj

**Comment by madhusudancs** [comment]
Having experienced the Kconfig pain, I have done some cursory search a few of times. I haven't found anything useful. How hard is it to write one?

**Comment by errordeveloper** [comment]
I was also thinking that a web app would be most useful.

**Comment by colemickens** [comment]
I actually didn't mind the use of Kconfig itself (as noted, it has a unique set of features that are useful here), but I struggled to get the `make` targets to depend on each other in such a way that I could opt-out of the Kconfig process if I'd already formed the json config that I wanted. For example, if the tool could be invoked once for "create config" and another for "do deployment from json config" then I think it would be fine. It was trying to chain Kconfig->json->deployment into a single command that made things more difficult in my case.

It would be neat if there was a tool that could consume a JSON Schema and walk the user through a CLI wizard that would process a conformant json config file as output. 

Actually, it looks like I found something that might be relevant... Angular + Json Schema = autogenerated HTML form for generating a JSON object... http://schemaform.io/examples/bootstrap-example.html

**Comment by colemickens** [comment]
This one actually supports oneOf which I think we would need for the cloud_provider specific bit of the config: http://jeremydorn.com/json-editor/

**Comment by cdornsife** [comment]
I have a fork that automates the .config creation using ENV vars and makes a .config file. The rest of the make runs normally.

Here is the readme:
https://github.com/applariat/kubernetes-anywhere/blob/master/AUTOMATED.md

Here is the python that makes the .config.
https://github.com/applariat/kubernetes-anywhere/blob/master/util/env_to_config.py

I had to change kconfiglib to recognize CONFIG_ env var since .config is prefixed with '.'
I submitted a PR to that project. If it is accepted then it can be pip installed. In the meantime I have an altered version of kconfiglib.py in utils.

If this looks useful I can create a PR.
